### PR TITLE
twitter support, finalizing

### PR DESCRIPTION
**Pull request by Misaka-0x447f**
| I   | T   |                             |                       |
| --- | --- | --------------------------- | --------------------- |
| x   |     | fetchPostContent            | not fully implemented, not testable |
| x   |     | fetchProfile                | not fully implemented, not testable |
| x   | x   | autoVerifyBio               | already working in prev stage
| x   | x   | autoVerifyPost              | already working in prev stage

- [x] user name collect error on bio card
- [x] auto paste post not work when post box already contains something
- [x] post collection not working on promoted post
- [x] saving friend status: friend, none (#330)
- [x] portrait view setup problem (#347)
## delayed to stage LTE:
[?] show welcome not detecting if current user has been setup (#193)
[?] use random website as prefix (#198) 
[?] fetchPostContent (facebook) not working properly (#323)
*LTE: Long term evolution

### CODE REVIEW

**Comment by neruthes** [comment]
Is this stage 5?

**Comment by Jack-Works** [comment]
> Is this stage 5?

image

**Comment by Misaka-0x447f** [comment]
Yes and this stage will not introduce new feature..

**Comment by neruthes** [comment]
What's up here?

**Comment by Misaka-0x447f** [comment]
this branch was automatically closed by github.

File: src/social-network-provider/twitter.com/ui/fetch.ts [hunk 1]
```diff
@@ -58,37 +67,40 @@ const registerUserCollector = () => {
...
                 })
                 // push to map. proxy used as a pointer here.
                 self.posts.set(proxy, info)
-            })
-            return {
-                onNodeMutation: collectPostInfo,
-                onTargetChanged: collectPostInfo,
-                onRemove: () => self.posts.delete(proxy),
-            }
+                return {
```
Comment (Jack-Works) [inline src/social-network-provider/twitter.com/ui/fetch.ts]: By this way, this function is not able to provide callbacks like `onNodeMutation`

File: src/social-network-provider/twitter.com/ui/tasks.ts [hunk 2]
```diff
@@ -1,20 +1,66 @@
...
-        console.warn('Text pasting failed')
+/**
+ * Wait for up to 5000 ms
+ * If not complete, let user do it.
+ */
+const taskPasteIntoPostBox: SocialNetworkUI['taskPasteIntoPostBox'] = (text, opt) => {
+    const interval = 500
+    const timeout = 5000
+    const worker = async function(abort: AbortController) {
+        const milestone = () => {
```
Comment (Jack-Works) [inline src/social-network-provider/twitter.com/ui/tasks.ts]: maybe a better name

File: src/social-network-provider/twitter.com/ui/tasks.ts [hunk 3]
```diff
@@ -1,20 +1,66 @@
...
+        milestone()
+        while (!hasFocus(i)) {
+            i.evaluate()!.focus()
+            milestone()
+            await sleep(interval)
+        }
+        // paste
+        dispatchCustomEvents('paste', text)
+        if (!getText().includes(text)) {
+            prompt(opt.warningText, text)
```
Comment (Jack-Works) [inline src/social-network-provider/twitter.com/ui/tasks.ts]: maybe throw here?

File: src/social-network-provider/twitter.com/utils/delegate.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+/**
+ * @desc
+ *  this file contains utils that used to
+ *  delegate some tasks to foreground by background.
+ *  we call foreground tabs 'worker' here.
+ */
+import { anyHostUrl, anyHostUrlMobile } from './url'
+import { shuffle } from 'lodash-es'
+
+export const workerAllocate = async () => {
```
Comment (Jack-Works) [inline src/social-network-provider/twitter.com/utils/delegate.ts]: I think the naming is confusing that the variables doesn't indicates what network does it serve. Like `anyHostUrl` is confusing in the global scope (auto importing). `anyHostURLForTwitter` will be better?

File: src/social-network/worker.ts [hunk 5]
```diff
@@ -12,6 +12,18 @@ import getCurrentNetworkWorker from './utils/getCurrentNetworkWorker'
 export interface SocialNetworkWorkerDefinition extends SocialNetworkWorkerAndUIDefinition {
     /**
      * This function should fetch the given post by `fetch`, `AutomatedTabTask` or anything
+     * @pseudoCode
+     * fetchPostContent(post) {
+     *      let tab = get_tab_with_same_origin_and_not_pinned()
```
Comment (Jack-Works) [inline src/social-network/worker.ts]: actually it doesn't have to be unpinned

File: src/utils/dom.ts [hunk 6]
```diff
@@ -1,13 +1,35 @@
-import { LiveSelector, MutationObserverWatcher } from '@holoflows/kit'
-import { timeout } from './utils'
+import { LiveSelector } from '@holoflows/kit'
+import { isUndefined } from 'lodash-es'
 
 export const untilElementAvailable = async (ls: LiveSelector<HTMLElement, boolean>) => {
-    await timeout(new MutationObserverWatcher(ls).startWatch({ childList: true, subtree: true }).then(), 10000)
+    return new Promise<void>((resolve, reject) => {
```
Comment (Jack-Works) [inline src/utils/dom.ts]: why change to this way to impl?

File: src/utils/utils.ts [hunk 7]
```diff
@@ -2,11 +2,46 @@ import { CustomEventId } from './constants'
...
+ *     do();
+ *     sleep(opt.interval)
+ * }
+ */
+const finiteWhileDoSleepDefault = { interval: 1000, timeout: 30000 }
+interface finiteWhileDoSleep {
+    interval?: number
+    timeout?: number
+}
+export const finiteWhileDoSleep = (
```
Comment (Jack-Works) [inline src/utils/utils.ts]: I didn't find where do you use this util. maybe remove it?

File: src/social-network-provider/twitter.com/ui/tasks.ts (hunk 2 above)
Comment (Misaka-0x447f) [reply src/social-network-provider/twitter.com/ui/tasks.ts]: 'checkSignal'

File: src/social-network-provider/twitter.com/ui/tasks.ts (hunk 3 above)
Comment (Misaka-0x447f) [reply src/social-network-provider/twitter.com/ui/tasks.ts]: done.

File: src/social-network-provider/twitter.com/utils/delegate.ts (hunk 4 above)
Comment (Misaka-0x447f) [reply src/social-network-provider/twitter.com/utils/delegate.ts]: renamed

File: src/utils/dom.ts (hunk 6 above)
Comment (Misaka-0x447f) [reply src/utils/dom.ts]: the prev implement was not working. maybe delay fixing?

File: src/utils/utils.ts (hunk 7 above)
Comment (Misaka-0x447f) [reply src/utils/utils.ts]: removed. it is designed to be used in auto paste.

File: src/social-network-provider/twitter.com/ui/fetch.ts (hunk 1 above)
Comment (Misaka-0x447f) [reply src/social-network-provider/twitter.com/ui/fetch.ts]: aa606d345c6c6bc5927fb9b32d01277ace497d05

File: src/social-network/worker.ts (hunk 5 above)
Comment (Misaka-0x447f) [reply src/social-network/worker.ts]: why? this is required when I tried to implement this...
or I cannot understand what it does.

File: src/utils/dom.ts (hunk 6 above)
Comment (Jack-Works) [reply src/utils/dom.ts]: maybe leave an issue for it

**Pull request merged**